API: Enforce one copy for ``__array__`` when ``copy=True``

mtsokol opened the pull request:
Addresses: https://github.com/numpy/numpy/issues/26208

Hi @seberg @ngoldbaum,

This PR enables passing `copy=True` to `def __array__(...)` and prevents NumPy from making another copy once it can assume that `copy=True` was successfully passed and the copy was made by the implementor. 

Following https://github.com/numpy/numpy/issues/26208#issuecomment-2037252734 I implemented it with a `copy_indicator` that is set to `1`, which later is used to decide whether `NPY_ARRAY_ENSURECOPY` flag can be removed. 

I added a test `test___array__copy_once` that checks if only one copy is made with `np.array(my_arr, copy=True)`. Without `copy_indicator` the test fails as `np.array(my_arr, copy=True)` doubles delta RSS (I used `psutil` package and added it as another test dependency).

I assumed that we want to prevent making implicit copies with `__array__`. Therefore the user is also responsible for properly implementing `dtype` argument.

```python
class MyArray:
    def __array__(self, dtype=None, copy=None):
        return np.array([1,2,3], dtype=int)

my_arr = MyArray()
np.array(my_arr, dtype=float)  # Error! Produced ndarray has incorrect dtype! `__array__` should use dtype arg
```

mtsokol committed:
API: Enforce one copy for __array__ when copy=True

seberg commented on numpy/_core/tests/test_multiarray.py:
```diff
@@ -8501,10 +8502,57 @@ def __array__(self, dtype=None):
+                else:
+                    return base_arr
+
+        process = psutil.Process()
```
Can you just do this without `psutils`:  Just store *copy* you return, then compare do an identity/base check.  Easier, and no need for psutil magic.

mtsokol replied:
Ah, right - I removed `psutils` and stored copies/base arrays instead.

seberg commented on numpy/_core/src/multiarray/array_coercion.c:
```diff
@@ -970,7 +972,7 @@ PyArray_DiscoverDTypeAndShape_Recursive(
         coercion_cache_obj ***coercion_cache_tail_ptr,
         PyArray_DTypeMeta *fixed_DType, enum _dtype_discovery_flags *flags,
-        int copy)
+        int copy, int *copy_indicator)
```
This already has the `_dtype_discovery_flags` I think it makes sense to use those "internally" to this part of the code.

mtsokol replied:
Do you mean including `copy` and `copy_indicator` arguments in `_dtype_discovery_flags`?

mtsokol replied:
Sure! I merged  `copy_indicator` into `_dtype_discovery_flags`.

seberg reviewed (commented):
Just two comments for now.  But yeah, we may need the copy indicator unfortunately (maybe call it `was_copied`, though?  It has to be passed by reference anyway, so it is obvious for the caller that it is an output).

It would be good to have a test that we don't pass `copy=True` if someone calls `np.array([array_like])` (i.e. there is a list).

mtsokol commented:
> (maybe call it `was_copied`, though?

@seberg Sure - renamed.

> It would be good to have a test that we don't pass `copy=True` if someone calls `np.array([array_like])` (i.e. there is a list).

Sure! I changed implementation so that sequences don't pass `copy=True` to it's elements and added a test for that.

mtsokol commented:
I did one more rebase to solve conflicts.

ngoldbaum reviewed (commented):
One request for clarification inline and one more general suggestion.

I think `was_copied` and `COPY_WAS_CREATED` should be renamed `was_copied_by__array__` and `COPY_WAS_CREATED_BY__ARRAY__` to emphasize that this is specifically to handle subtleties around `__array__` implementations. Alternative names or just comments around where they are used explaining that it has something to do with `__array__` would help readers trying to understand why we have both an ensure copy flag and a flag tracking if a copy already happened.

mtsokol committed:
Adjust was_copied variable names

charris commented:
Does this need a backport?

ngoldbaum commented:
> Does this need a backport?

No this change is targeted for 2.1

mtsokol commented:
> I think `was_copied` and `COPY_WAS_CREATED` should be renamed `was_copied_by__array__` and `COPY_WAS_CREATED_BY__ARRAY__` to emphasize that this is specifically to handle subtleties around `__array__` implementations. 

These names are clearer and more specific - done!

seberg commented on numpy/_core/src/multiarray/ctors.c:
```diff
@@ -2577,10 +2590,12 @@ PyArray_FromArrayAttr_int(
-        /* TODO: As of NumPy 2.0 this path is only reachable by C-API. */
-        Py_SETREF(new, PyArray_NewCopy((PyArrayObject *)new, NPY_KEEPORDER));
+    if (was_copied_by__array__ != NULL && copy == 1 &&
+        must_copy_but_copy_kwarg_unimplemented == 0) {
```
Could add a comment to remove the `was_copied_by__array__` argument again here, but it is probably clear enough (when it happens).

mtsokol replied:
Done!

seberg commented on numpy/_core/src/multiarray/ctors.c:
```diff
@@ -1615,6 +1622,10 @@ PyArray_FromAny_int(PyObject *op, PyArray_Descr *in_descr,
         assert(cache->converted_obj == op);
         arr = (PyArrayObject *)(cache->arr_or_sequence);
         /* we may need to cast or assert flags (e.g. copy) */
+        if (was_copied_by__array__ == 1 && flags & NPY_ARRAY_ENSURECOPY) {
```
```suggestion
        if (was_copied_by__array__ == 1) {
```
You don't need to test this, it is OK to unset the flag if it isn't set.  (to me that is clearer, but maybe it's me)

mtsokol replied:
Done!

seberg commented on numpy/_core/src/multiarray/ctors.c:
```diff
@@ -1615,6 +1622,10 @@ PyArray_FromAny_int(PyObject *op, PyArray_Descr *in_descr,
         /* we may need to cast or assert flags (e.g. copy) */
+        if (was_copied_by__array__ == 1 && flags & NPY_ARRAY_ENSURECOPY) {
+            flags = flags & ~NPY_ARRAY_ENSURECOPY;
+            flags = flags | NPY_ARRAY_ENSURENOCOPY;
```
This looks wrong, so please add a test:
```
np.asarray(array_like, copy=True, order="F")
```
should just make two copies.  You probably want to make sure that users honors the `dtype` that was passed.  That might be nice, but if it is not very clean to test for, I am not sure we should worry much.

mtsokol replied:
Ah right, a copy still needs to be made to convert to `order="F"`. Then I shouldn't add `ENSURENOCOPY` flag but only remove `ENSURECOPY` one.

About honoring `dtype`: I can check if `arr` in cache has the same `dtype` as requested `dtype` variable. I wonder if it's sufficient for comparing dtypes in C-API:
```
PyArray_TYPE(arr) == dtype->type_num
```

mtsokol replied:
Then let's maybe skip checking `dtype` after `__array__` call? 
What we currently have is that:
```python
np.arrray(array_like, dtype=new_dtype, copy=False)
```
errors when a `dtype` wasn't honored in `__array__` implementation and
```python
np.arrray(array_like, dtype=new_dtype)
```
get's another copy, same as `order="F"` case.

seberg replied:
> Then let's maybe skip checking dtype after __array__ call?
For here, that is perfectly fine to me, it is just a mild performance nudge for downstream after all.

However, maybe we should make a new issue.  I am slightly worried that downstream until now always just ignored `dtype=`, and now we will get new bugs with subarray dtypes.

OTOH, I will also say that, I don't expect anyone to explicitly run into it, so in a sense, it is an unrelated bug fix.

mtsokol replied:
I updated flags logic here and added a test with:
```python
np.asarray(array_like, copy=True, order="F")
```

seberg reviewed (commented):
LGTM, I suppose we could use the argument in `PyArray_DiscoverDTypeAndShape` for other things in principle, although I am not sure there is anything concrete, so let's use the name, it is much nicer everywhere else (and all other places will go away when the deprecation goes away).

The `ENSURE_COPY` flag setting logic looks incorrect, though (I admit, I didn't try it).

mtsokol committed:
Apply review comments

mtsokol committed:
linting

ngoldbaum reviewed (approved):
I think this is OK to merge now.

I don't think this is going to cause much additional churn over adding the `copy` keyword in the first place, hopefully downstream did the right thing.

mtsokol commented:
Ok, merging it now.
